library/persistence.js

```diff
@@ -0,0 +1,138 @@
+const { MongoClient } = require('mongodb')
+
+function connect (url = 'mongodb://localhost:27017', dbName = 'pinano', options = {}) {
+  options.useNewUrlParser = true
+  return MongoClient.connect(url, options).then(client => {
+    return new MongoManager(client, dbName)
+  })
+}
+
+class MongoManager {
+  constructor (client, dbName) {
+    this.client = client
+    this.dbName = dbName
+    this.db = this.client.db(this.dbName)
+  }
+
+  dropDatabase () {
+    return this.db.dropDatabase()
+  }
+
+  newUserRepository () {
+    return new MongoUserRepository({ userCollection: this.db.collection('users') })
+  }
+
+  newGuildRepository () {
+    return new MongoGuildRepository({ guildCollection: this.db.collection('guilds') })
+  }
+
+  async shutdown () {
+    await this.client.close()
+  }
+}
+
+function makeUser (userId) {
+  return {
+    id: userId,
+    current_session_playtime: 0,
+    overall_session_playtime: 0
+  }
+}
+
+class MongoUserRepository {
+  constructor ({ userCollection }) {
+    this.collection = userCollection
+  }
+
+  async load (userId) {
+    return this.collection.findOne({ id: userId })
+  }
+
+  async loadTopSession (n) {
+    return this._loadTopBy(n, 'current_session_playtime')
+  }
+
+  async loadTopOverall (n) {
+    return this._loadTopBy(n, 'overall_session_playtime')
+  }
+
+  async _loadTopBy (n, key) {
+    return this.collection.aggregate([
```

Comment:
Letting the db do the work for us. Nice.

Reply:
It does occur to me, though, that this data may not (will not?) take liveness into account.

Reply:
Hmmm.

One idea I had (which would also help with people losing time during deploys/crashes) would be to have us monitor everyone active and write their updated session time to the database once a minute or something like that. That would help sort of with this problem.